lib/src/ssh_signing.rs

```diff
@@ -0,0 +1,305 @@
+
+use std::ffi::OsString;
+use std::fmt::Debug;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::{Command, ExitStatus, Stdio};
+
+use thiserror::Error;
+
+use crate::signing::{SigStatus, SignError, SigningBackend, Verification};
+
+#[derive(Debug)]
+pub struct SshBackend {
+    program: OsString,
+    allowed_signers: Option<OsString>,
+}
+
+#[derive(Debug, Error)]
+pub enum SshError {
+    #[error("SSH sign failed with exit status {exit_status}:\n{stderr}")]
+    Command {
+        exit_status: ExitStatus,
+        stderr: String,
+    },
+    #[error("Failed to parse ssh program response")]
+    BadResult,
+    #[error("Failed to run ssh-keygen")]
+    Io(#[from] std::io::Error),
+    #[error("Signing key required")]
+    MissingKey,
+    #[error("Allowed signers file not provided")]
+    MissingAllowedSigners,
+}
+
+impl From<SshError> for SignError {
+    fn from(e: SshError) -> Self {
+        SignError::Backend(Box::new(e))
+    }
+}
+
+type SshResult<T> = Result<T, SshError>;
+
+fn parse_utf8_string(data: Vec<u8>) -> SshResult<String> {
+    String::from_utf8(data).map_err(|_| SshError::BadResult)
+}
+
+fn run_command(command: &mut Command, stdin: &[u8]) -> SshResult<Vec<u8>> {
+    let process = command.spawn()?;
+    process.stdin.as_ref().unwrap().write_all(stdin)?;
+
+    let output = process.wait_with_output()?;
+
+    if !output.status.success() {
+        return Err(SshError::Command {
+            exit_status: output.status,
+            stderr: String::from_utf8_lossy(&output.stderr).into(),
+        });
+    }
+
+    Ok(output.stdout)
+}
+
+// This attempts to convert given key data into a file and return the filepath.
+// If the given data is actually already a filepath to an existing file on disk
+// then the key input is returned directly.
+fn ensure_key_as_file(key: &str) -> SshResult<(PathBuf, Option<tempfile::TempPath>)> {
+    let key_path = Path::new(key);
+    if key_path.try_exists()? {
+        println!("Yea returning here");
+        return Ok((key_path.to_path_buf(), None));
+    }
+
+    // The ssh-keygen `-f` flag expects to be given a file which contains either a
+    // private or public key.
+    //
+    // As it expects a file and we will generally have just public key data we need
+    // to put it into a file first.
+    let mut pub_key_file = tempfile::Builder::new()
+        .prefix("jj-signing-key-")
+        .tempfile()
+        .map_err(SshError::Io)?;
+
+    pub_key_file
+        .write_all(key.as_bytes())
+        .map_err(SshError::Io)?;
+    pub_key_file.flush().map_err(SshError::Io)?;
+
+    // This is converted into a TempPath so that the underlying file handle is
+    // closed. On Windows systems this is required for other programs to be able
+    // to open the file for reading.
+    let pub_key_path = pub_key_file.into_temp_path();
+    Ok((pub_key_path.to_path_buf(), Some(pub_key_path)))
+}
+
+#[test]
+fn test_ssh_key_to_file_conversion_raw_key_data() {
```

Comment:
I actually don't remember even seeing tests like these (unit-ish directly next to source) in jj, but anyway in Rust usually such tests are put in

```rust
#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn my_test() {
    assert_eq!(2 + 2, 4)
  }
}
```
module at the bottom of a source file